compiler/rustc_hir_analysis/src/check/check.rs

```diff
@@ -121,7 +121,7 @@ fn check_union_fields(tcx: TyCtxt<'_>, span: Span, item_def_id: LocalDefId) -> b
 
         let param_env = tcx.param_env(item_def_id);
         for field in &def.non_enum_variant().fields {
-            let field_ty = field.ty(tcx, substs);
+            let field_ty = tcx.normalize_erasing_regions(param_env, field.ty(tcx, substs));
```

Comment:
this would ICE if the union has fields which aren't well formed, e.g. something like `<T as NotImplemented>::Assoc`. Please add a test for this and check whether it currently causes an ICE. If so you can replace this with `try_normalize_erasing_regions`. I gues it's fine for normalization to ignore region obligations as we should detect these region errors in WF check.

While this also ignore

Reply:
not sure when exactly `normalize_erasing_regions` ends up failing with an ICE, only know that it does. Maybe also from something like

```rust
trait Identity {
    type Identity;
}
trait NotImplemented {}

impl<T: NotImplemented> Identity for T {
    type Identity = Self;
}

type Foo = u8;

union Bar {
    a:  <Foo as Identity>::Identity,
    b: u8,
}
```

Reply:
Working too, adding it as a test.

Reply:
> If there's a way to ICE here we will just get a bug report, so that seems fine

#106938 :rofl:

Reply:
dis you mean #107872 ?

Reply:
🤦 yes

Reply:
That didn't take long. :rofl:

Reply:
Showed up as new in my daily ICE-screening ^^

Reply:
Great job!